File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/sensors/optical/AmbientLightSensor.java
package io.treehopper.libraries.sensors.optical;

public abstract class AmbientLightSensor implements IAmbientLightSensor {
    protected double lux;
    private boolean autoUpdateWhenPropertyRead = true;

    @Override
    public double getLux() {
        if (isAutoUpdateWhenPropertyRead())
            update();

        return lux;
    }

    @Override
    public boolean isAutoUpdateWhenPropertyRead() {
        return autoUpdateWhenPropertyRead;
    }

    @Override
    public void setAutoUpdateWhenPropertyRead(boolean value) {
        autoUpdateWhenPropertyRead = value;
    }
}

File: C++/API/src/libraries/Sensors/Optical/AmbientLightSensor.cpp
#include <Libraries/Sensors/Optical/AmbientLightSensor.h>

namespace Treehopper {
    namespace Libraries {
        namespace Sensors {
            namespace Optical {
                double AmbientLightSensor::lux() {
                    if (autoUpdateWhenPropertyRead) update();

                    return _lux;
                }
            }
        }
    }
}
